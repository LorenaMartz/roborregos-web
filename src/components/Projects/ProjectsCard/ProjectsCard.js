// @flow
import React, { Component } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleDown } from '@fortawesome/free-solid-svg-icons'
import { Row, Col, Container } from 'react-bootstrap'
import logo from '../../../images/info_background_3.jpg'
import './ProjectsCard.css'

type Props = {};

/** Component class of Projects' card. */
class ProjectsCard extends Component<Props> {
  joinUsCallback() {
    window.open('https://www.facebook.com/RoBorregos/')
  }

  /**
 * Renders Responsive view of Projects's card.
 * @return {renderized_component} Heder banner with legend.
 */
  render() {
    return (
      <div className="projects-card">
        <Row>
          <Col lg="5">
            <img src={logo} className="card-image" alt="logo" />
          </Col>
          <Col lg="7" className="card-info">
            <h2 className="title-text-card">
              @Home
            </h2>
            <div className="main-text-projects">
              <p>
                We are designing, developing, and manufacturing a home assistance robot that can carry out activities for the safety and comfort of the user, such as transporting things from different places in the home to seeing who is at the door of the house and welcoming them! It’s based on the “Robocup @Home” competition.
              </p>
            </div>
            <button className="card-button" onClick={this.joinUsCallback} variant="outline-primary">
              Learn more
            </button>
          </Col>
        </Row>
        {/* <div className="container-legend">
          <h2 className="title-text-card">
            Projects
          </h2>
          <div className="main-text-projects">
            <p>
              Every year we challenge ourselves to keep learning and improving.
              {' '}
              <br />
              See what we&apos;ve been up to!
            </p>
          </div>
        </div> */}
      </div>
    )
  }
}

export default ProjectsCard
