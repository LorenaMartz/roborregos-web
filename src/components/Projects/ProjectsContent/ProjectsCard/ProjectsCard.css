.projects-card {
    height: 90vh;
    position: relative;
    background: black;
    color: rgba(255, 255, 255, 0.8);
  }

.card-info {
    padding: 10vw;
}

.card-button {
    margin: auto;
    cursor: pointer;
    border: 1px solid #4F72B8;
    border-radius: 8px !important;
    width: 15vw;
    height: 60px;
    background-color: #13191E;
    color: white !important;
    transition: 0.3s;
    font-size: 1.5vw;
}

.card-image {
    display: block;
    object-fit: cover;
    height: 100%;
    max-width: 100%;
}

.projects-card .main-text-projects {
    color: rgba(255, 255, 255, 0.9);
    font-size: 3vh;
    text-transform: none;
    padding-top: 2%;
    padding-bottom: 3vh;
}

.title-text-card {
    color: rgba(255, 255, 255, 0.9);
    font-size: 7vh;
    text-transform: none;
}

@media (max-width: 768px) {
    .title-text-card {
        text-align: center;
        height: 15vh;
        padding-top: 5vh;
    }

    .main-text-projects-description {
        height: 30vh;
        padding-left: 6vw;
        padding-right: 6vw;
        padding-top: 4vh;
        text-align: center;
        font-size: 100%;
    }

    .card-image-container {
        height: 35vh;
        width: 100vw;
    }

    .card-image {
        width: 100%;
        height: 100%;
    }

    .button-div {
        height: 8vh;
    }

    .card-button-small {
        font-size: 4vw;
        width: 40vw;
        display: block;
        margin-left: auto;
        margin-right: auto;
        padding-left: 5vw;
        padding-right: 5vw;
        border: 1px solid #4F72B8;
        border-radius: 8px !important;
        background-color: #13191E;
        color: white !important;
        transition: 0.3s;
    }
}
