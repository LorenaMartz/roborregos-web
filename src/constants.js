export const LARGE_WIDTH = 1200
export const MEDIUM_WIDTH = 992
export const SMALL_WIDTH = 768
export const XSMALL_WIDTH = 592
export const MOBILE_WIDTH = 414
