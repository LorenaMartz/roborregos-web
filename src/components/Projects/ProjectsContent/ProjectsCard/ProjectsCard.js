// @flow
import React, { Component } from 'react'
import { Row, Col } from 'react-bootstrap'
import './ProjectsCard.css'
import Fade from '@material-ui/core/Fade'
import Grow from '@material-ui/core/Grow'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleDown } from '@fortawesome/free-solid-svg-icons'
import { SMALL_WIDTH } from '../../../../constants'

type Project = {
  title: string,
  description: string,
  image: string,
  wiki: string
};

type Props = {
  index: number,
  project: Project,
  show_scrollers: boolean
};

type State = {
  large_view: boolean
};

/** Component class of Projects' card. */
class ProjectsCard extends Component<Props, State> {
  project: Project;

  index: number;

  show_scrollers: boolean;

  /**
   * Class constructor
   * @param {list} props: List of projects data.
   */
  constructor(props: Props) {
    super(props)
    this.cardContent = this.cardContent.bind(this)
    this.imageContent = this.imageContent.bind(this)
    this.largeView = this.largeView.bind(this)
    this.viewSizeLarge = this.viewSizeLarge.bind(this)
    this.smallView = this.smallView.bind(this)

    this.project = props.project
    this.index = props.index
    this.show_scrollers = props.show_scrollers
    this.state = {
      large_view: this.viewSizeLarge(),
    }
  }

  componentDidMount() {
    window.addEventListener('resize', this.updateView)
  }

  updateView = () => {
    this.setState({
      large_view: this.viewSizeLarge(),
    })
  }

  viewSizeLarge = () => {
    if (window.innerWidth > SMALL_WIDTH) {
      return true
    }
    return false
  }

  projectCallback = (link: string) => {
    window.open(link)
  }

  /**
  * Parses path to projects image
  * @param {string} imgPath: Path to project's image/photo.
  * @return {path}
  */
  tryRequire = (imgPath: string) => {
    try {
      /* eslint-disable import/no-dynamic-require */
      /* eslint-disable global-require */
      return require(`../../../../images/projects/${imgPath}`)
    } catch (err) {
      return imgPath
    }
  }

  cardContent = () => (
    <Grow in {...{ timeout: 1500 }} style={{ transformOrigin: 'bottom' }}>
      <Col xs={6} className="card-info">
        <h2 className="title-text-card">
          {this.project.title}
        </h2>
        <div className="main-text-projects">
          <p>
            {this.project.description}
          </p>
        </div>
        <button type="button" className="card-button" onClick={() => { if (this.project.wiki !== '') this.projectCallback(this.project.wiki) }} variant="outline-primary">
          {(this.project.wiki !== '') ? 'Learn more' : 'Coming soon'}
        </button>
      </Col>
    </Grow>
  )

  largeView = () => {
    if (this.index % 2) {
      return (
        <Row className="projects-card">
          {this.cardContent()}
          {this.imageContent()}
        </Row>
      )
    }
    return (
      <Row
        className="projects-card"
        style={{ paddingLeft: '5vw' }}
      >
        {this.imageContent()}
        {this.cardContent()}
      </Row>
    )
  }

  getScrollers = () => {
    if (this.show_scrollers) {
      return (
        <div className="scroll-next">
          <FontAwesomeIcon
            onClick={() => this.scrollToInfo()}
            icon={faAngleDown}
            className="scroll-next-icon"
          />
        </div>
      )
    }
    return (<div />)
  }

  scrollToInfo = () => {
    window.scrollBy(0, window.innerHeight * (1 - 0.05)
    - (window.scrollY - (this.index + 1) * window.innerHeight * (1 - 0.05)))
  }

  smallView = () => (
    <div className="projects-card">
      <div className="title-text-card">
        <h2>
          {this.project.title}
        </h2>
      </div>
      <div className="card-image-container">
        <img className="card-image" src={this.tryRequire(`${this.project.image}.jpg`)} alt={this.project.title} />
      </div>
      <div className="main-text-projects-description">
        {this.project.description}
      </div>
      <div className="button-div">
        <button
          type="button"
          className="card-button-small"
          onClick={() => { if (this.project.wiki !== '') this.projectCallback(this.project.wiki) }}
          variant="outline-primary"
        >
          {(this.project.wiki !== '') ? 'Learn more' : 'Coming soon'}
        </button>
      </div>
    </div>
  )

  imageContent = () => (
    <Col xs={6}>
      <img src={this.tryRequire(`${this.project.image}.jpg`)} className="card-image" alt={this.project.title} />
    </Col>
  )

  /**
 * Renders Responsive view of Projects's card.
 * @return {renderized_component} Heder banner with legend.
 */
  render() {
    const { large_view } = this.state
    if (large_view) {
      return (
        <div id={`card-${this.index}`}>
          <Fade in {...{ timeout: 2000 }}>
            {this.largeView()}
          </Fade>
          {this.getScrollers()}
        </div>
      )
    }
    return this.smallView()
  }
}

export default ProjectsCard
