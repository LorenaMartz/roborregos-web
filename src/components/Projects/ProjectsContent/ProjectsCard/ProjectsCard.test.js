// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, configure, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import ProjectsCard from './ProjectsCard'
import { SMALL_WIDTH, LARGE_WIDTH } from '../../../../constants'

const project = {
  title: 'project-1',
  description: 'description-1',
  wiki: 'https://github.com/1',
  image: 'project-1',
}

const project_in_progress = {
  title: 'project-1',
  description: 'description-1',
  wiki: '',
  image: 'project-1',
}

const index = 0

let wrapper
configure({ adapter: new Adapter() })

beforeEach(() => {
  global.innerWidth = LARGE_WIDTH
  global.dispatchEvent(new Event('resize'))
  wrapper = shallow(
    <ProjectsCard project={project} index={index} show_scrollers={false} />,
  )
})

it('<ProjectCard> large view renders components correctly', () => {
  expect(wrapper.find('.projects-card').length).toBe(1)
  expect(wrapper.find('.card-info').length).toBe(1)
  expect(wrapper.find('.title-text-card').length).toBe(1)
  expect(wrapper.find('.main-text-projects').length).toBe(1)
  expect(wrapper.find('.card-image').length).toBe(1)
  expect(wrapper.find('.card-button').length).toBe(1)
  const button_text = wrapper.find('.card-button').text()
  expect(button_text).toEqual('Learn more')
})

it('<ProjectCard> small view renders components correctly', () => {
  global.innerWidth = SMALL_WIDTH
  global.dispatchEvent(new Event('resize'))
  expect(wrapper.find('.projects-card').length).toBe(1)
  expect(wrapper.find('.title-text-card').length).toBe(1)
  expect(wrapper.find('.card-image-container').length).toBe(1)
  expect(wrapper.find('.main-text-projects-description').length).toBe(1)
  expect(wrapper.find('.button-div').length).toBe(1)
  expect(wrapper.find('.card-button-small').length).toBe(1)
  expect(wrapper.find('.card-button-small').text()).toEqual('Learn more')
})

it('<ProjectCard> renders button text correctly if given a project in progress (no wiki)', () => {
  wrapper = shallow(
    <ProjectsCard project={project_in_progress} index={index} show_scrollers={false} />,
  )
  expect(wrapper.find('.card-button').text()).toEqual('Coming soon')
})

it('<ProjectCard> opens link when button is clicked in large view', () => {
  global.innerWidth = LARGE_WIDTH
  global.dispatchEvent(new Event('resize'))
  global.open = jest.fn()
  wrapper.find('.card-button').simulate('click')
  expect(global.open).toHaveBeenCalledWith(project.wiki)
})

it('<ProjectCard> opens link when button is clicked in small view', () => {
  global.innerWidth = SMALL_WIDTH
  global.dispatchEvent(new Event('resize'))
  global.open = jest.fn()
  wrapper.find('.card-button-small').simulate('click')
  expect(global.open).toHaveBeenCalledWith(project.wiki)
})

it('<ProjectCard> renders scroller in large view if flagged as prop', () => {
  wrapper = shallow(
    <ProjectsCard project={project_in_progress} index={index} show_scrollers />,
  )
  expect(wrapper.find('.icon-container').length).toBe(1)
})

it('<ProjectCard> calls scrollTo when blue arrow is clicked in large view', () => {
  global.innerWidth = LARGE_WIDTH
  global.dispatchEvent(new Event('resize'))
  wrapper = shallow(
    <ProjectsCard project={project_in_progress} index={index} show_scrollers />,
  )
  global.scrollBy = jest.fn()
  wrapper.find('.icon-btn').simulate('click')
  expect(global.scrollBy).toHaveBeenCalledWith(
    0, window.innerHeight * (index + 2) - (85 * (index + 1))
  - window.scrollY + 35 * index,
  )
})

// Snapshots
describe('<ProjectCard> matches snapshot:', () => {
  it('Large width', () => {
    global.innerWidth = LARGE_WIDTH
    global.dispatchEvent(new Event('resize'))
    const wrapper_content = mount(
      <ProjectsCard project={project} index={0} show_scrollers={false} />,
    ).debug()
    expect(wrapper_content).toMatchSnapshot('ProjectCardLargeView.test.js.snap')
  })

  it('Small width', () => {
    global.innerWidth = SMALL_WIDTH
    global.dispatchEvent(new Event('resize'))
    const tree = renderer.create(
      <ProjectsCard project={project} index={0} show_scrollers={false} />,
    ).toJSON()
    expect(tree).toMatchSnapshot('ProjectCardSmallView.test.js.snap')
  })
})
