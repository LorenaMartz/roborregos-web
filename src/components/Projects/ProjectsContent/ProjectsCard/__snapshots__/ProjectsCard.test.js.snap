// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectCard> matches snapshot: Large width: ProjectCardLargeView.test.js.snap 1`] = `
"<ProjectsCard project={{...}} index={0} show_scrollers={false}>
  <div>
    <ForwardRef(Fade) in={true} timeout={2000}>
      <Transition appear={true} in={true} nodeRef={[undefined]} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExited={[Function (anonymous)]} onExiting={[Function (anonymous)]} timeout={2000} mountOnEnter={false} unmountOnExit={false} enter={true} exit={true}>
        <Row className=\\"projects-card\\" style={{...}} noGutters={false}>
          <div style={{...}} className=\\"projects-card row\\">
            <Col xs={6}>
              <div className=\\"col-6\\">
                <img src=\\"project-1.jpg\\" className=\\"card-image\\" alt=\\"project-1\\" />
              </div>
            </Col>
            <ForwardRef(Grow) in={true} timeout={1500} style={{...}}>
              <Transition appear={true} in={true} nodeRef={[undefined]} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} onEntering={[Function (anonymous)]} onExit={[Function (anonymous)]} onExited={[Function (anonymous)]} onExiting={[Function (anonymous)]} addEndListener={[Function: addEndListener]} timeout={1500} mountOnEnter={false} unmountOnExit={false} enter={true} exit={true}>
                <Col xs={6} className=\\"card-info\\" style={{...}}>
                  <div style={{...}} className=\\"card-info col-6\\">
                    <h2 className=\\"title-text-card\\">
                      project-1
                    </h2>
                    <div className=\\"main-text-projects\\">
                      <p>
                        description-1
                      </p>
                    </div>
                    <button type=\\"button\\" className=\\"card-button\\" onClick={[Function: onClick]} variant=\\"outline-primary\\">
                      Learn more
                    </button>
                  </div>
                </Col>
              </Transition>
            </ForwardRef(Grow)>
          </div>
        </Row>
      </Transition>
    </ForwardRef(Fade)>
    <div />
  </div>
</ProjectsCard>"
`;

exports[`<ProjectCard> matches snapshot: Small width: ProjectCardSmallView.test.js.snap 1`] = `
<div
  className="projects-card"
>
  <div
    className="title-text-card"
  >
    <h2>
      project-1
    </h2>
  </div>
  <div
    className="card-image-container"
  >
    <img
      alt="project-1"
      className="card-image"
      src="project-1.jpg"
    />
  </div>
  <div
    className="main-text-projects-description"
  >
    description-1
  </div>
  <div
    className="button-div"
  >
    <button
      className="card-button-small"
      onClick={[Function]}
      type="button"
      variant="outline-primary"
    >
      Learn more
    </button>
  </div>
</div>
`;
